libs/s25main/mapGenerator/Reshaper.cpp

```diff
@@ -0,0 +1,215 @@
+    
+    const double maxDistance = GridDistance(zero, center, size);
+    
+    double alpha;
+    
+    for (int x = 0; x < size.x; ++x)
+    {
+        for (int y = 0; y < size.y; ++y)
+        {
+            alpha = 1.0 - GridDistance(Position(x,y), center, size) / maxDistance;
+            z[x + y * size.x] += Scale(alpha, scale);
+        }
+    }
+}
+
+void Reshaper::ElevateEdges(HeightMap& z, const MapExtent& size, double scale)
+{
+    double alpha;
+    int dx, dy;
+
+    for (int x = 0; x < size.x; ++x)
+    {
+        for (int y = 0; y < size.y; ++y)
+        {
+            dx = std::min(x, size.x-x-1);
+            dy = std::min(y, size.y-y-1);
+            
+            if (dx < dy)
+            {
+                alpha = 1.0 - 2 * static_cast<double>(dx) / static_cast<double>(size.x);
+            }
+            else
+            {
+                alpha = 1.0 - 2 * static_cast<double>(dy) / static_cast<double>(size.y);
+            }
+            
+            z[x + y * size.x] += Scale(alpha, scale);
+        }
+    }
+}
+
+void Reshaper::ElevateCorners(HeightMap& z, const MapExtent& size, double scale)
+{
+    const Position zero(0,0);
+    const Position center(size/2);
+    
+    const double maxDistance = GridDistance(zero, center, size);
+    
+    double alpha;
+    
+    for (int x = 0; x < size.x; ++x)
+    {
+        for (int y = 0; y < size.y; ++y)
+        {
+            alpha = GridDistance(Position(x,y), center, size) / maxDistance;
+            z[x + y * size.x] += Scale(alpha, scale);
+        }
+    }
+}
+
+void Reshaper::ElevateContrast(HeightMap& z, const MapExtent& size, double scale)
+{
+    const int nodes = size.x * size.y;
+
+    const auto mean = std::accumulate(z.begin(), z.end(), 0u) / nodes;
+    const auto maximum = *std::max_element(z.begin(), z.end());
+    
+    for (int i = 0; i < nodes; ++i)
+    {
+        z[i] += Scale(1.0 - abs(static_cast<int>(z[i] - mean)) / static_cast<double>(maximum), scale);
+    }
+}
+
+void Reshaper::ElevateRandom(HeightMap& z, const MapExtent& size, double scale)
+{
+    const int nodes = size.x * size.y;
+    
+    for (int i = 0; i < nodes; ++i)
+    {
+        z[i] += Scale(0.5, scale);
+    }
+}
+
+void Reshaper::Elevate(HeightMap& z, ElevationMode mode, const MapExtent& size, double scale)
+{
+    switch (mode)
+    {
+        case Center:
+            ElevateCenter(z, size, scale);
+            break;
+        
+        case Edges:
+            ElevateEdges(z, size, scale);
+            break;
+        
+        case Corners:
+            ElevateCorners(z, size, scale);
+            break;
+            
+        case Contrast:
+            ElevateContrast(z, size, scale);
+            break;
+            
+        case Random:
+            ElevateRandom(z, size, scale);
+            break;
+            
+        default:
+            break;
+    }
+}
+
+double Reshaper::ScaleCorner(int x, int y, ElevationCorner corner, const MapExtent& size) const
```

Comment:
Please avoid taking a `x,y` pair and take a `Position` instead. Same performance, better semantics and readability.